File: RailgunVB/Commands/WhoIs.vb
Imports System.Text
Imports Discord
Imports Discord.Commands

Namespace Commands
    
    <Group("whois")>
    Public Class WhoIs
        Inherits ModuleBase
        
        <Command>
        Public Async Function WhoisAsync(userId As ULong) As Task
            Dim user As IUser = Await Context.Client.GetUserAsync(userId)
            
            If user Is Nothing
                Await ReplyAsync("Can not find user.")
                Return
            End If
            
            Await WhoisAsync(user)
        End Function
        
        Public Async Function WhoisAsync(Optional pUser As IUser = Nothing) As Task
            Dim user As IUser = If(pUser Is Nothing, Context.User, pUser)
            Dim username As String = $"{user.Username}#{user.DiscriminatorValue}"
            Dim isHuman As String = If(user.IsBot Or user.IsWebhook, "Bot", "User")
            Dim builder As New EmbedBuilder With {
                    .Title = $"Who is {Format.Bold(username)}?",
                    .Color = Color.Blue,
                    .ThumbnailUrl = user.GetAvatarUrl(),
                    .Footer = new EmbedFooterBuilder() With {
                        .Text = $"User ID : {user.Id}"
                    }
                }
            
            builder.AddField("User Type:", isHuman, true) _
                .AddField("Registered:", user.CreatedAt, true)
            
            Dim gUser As IGuildUser = Await Context.Guild.GetUserAsync(user.Id)
            
            If gUser IsNot Nothing
                Dim roles As New StringBuilder
                Dim esperLevel As Integer = GetEsperLevel(gUser)
                
                For Each roleId As ULong In gUser.RoleIds
                    roles.AppendFormat("| {0} |", Context.Guild.GetRole(roleId).Mention)
                Next
                
                builder.AddField("Server Nickname:", If(gUser.Nickname, "N/A"), true) _
                    .AddField("Joined Server At:", gUser.JoinedAt, true) _
                    .AddField("Current Server Roles:", If(roles.ToString(), "N/A")) _
                    .AddField("ESPer Level:", esperLevel)
            End If
            
            Await ReplyAsync("", embed := builder.Build())
        End Function
        
        Private Function GetEsperLevel(gUser As IGuildUser) As Integer
            If Context.Guild.OwnerId = gUser.Id Then Return 6
            
            Dim score = 0
            
            For Each perm As GuildPermission In gUser.GuildPermissions.ToList()
                Select perm
                    Case GuildPermission.Administrator
                        score += 214
                        Exit Select
                    Case GuildPermission.ManageChannels
                    Case GuildPermission.ManageEmojis
                    Case GuildPermission.ManageGuild
                    Case GuildPermission.ManageRoles
                    Case GuildPermission.ManageWebhooks
                    Case GuildPermission.ViewAuditLog
                        score += 10
                        Exit Select
                    Case GuildPermission.BanMembers
                    Case GuildPermission.DeafenMembers
                    Case GuildPermission.KickMembers
                    Case GuildPermission.MoveMembers
                    Case GuildPermission.MuteMembers
                    Case GuildPermission.ManageMessages
                    Case GuildPermission.ManageNicknames
                    Case GuildPermission.PrioritySpeaker
                    Case GuildPermission.MentionEveryone
                        score += 5
                        Exit Select
                    Case GuildPermission.AddReactions
                    Case GuildPermission.AttachFiles
                    Case GuildPermission.ChangeNickname
                    Case GuildPermission.CreateInstantInvite
                    Case GuildPermission.EmbedLinks
                    Case GuildPermission.ReadMessageHistory
                    Case GuildPermission.SendMessages
                    Case GuildPermission.SendTTSMessages
                    Case GuildPermission.UseExternalEmojis
                    Case GuildPermission.ViewChannel
                    Case GuildPermission.Connect
                    Case GuildPermission.Speak
                    Case GuildPermission.UseVAD
                        score += 1
                        Exit Select
                End Select
            Next
            
            Dim percent As Integer = Math.Round((score / 214) * 100)
            
            Select percent
                Case >= 100
                    Return 5
                Case >= 80
                    Return 4
                Case >= 60
                    Return 3
                Case >= 40
                    Return 2
                Case >= 20
                    Return 1
                Case Else
                    Return 0
            End Select
        End Function
    
    End Class
    
End NameSpace

File: Railgun/Commands/WhoIs.cs
using System;
using System.Text;
using System.Threading.Tasks;
using Discord;
using Finite.Commands;
using Railgun.Core;

namespace Railgun.Commands
{
    [Alias("whois")]
    public class WhoIs : SystemBase
    {
        private (double Percent, int Level, int PermissionCount) GetEsperLevel(IGuildUser user) 
        {
            var score = 0;
            var permList = user.GuildPermissions.ToList();
            
            if (Context.Guild.OwnerId == user.Id) return (-1, 6, permList.Count);
            if (permList.Contains(GuildPermission.Administrator)) return (-1, 5, permList.Count);

            foreach (var perm in permList) 
            {
                switch (perm) 
                {
                    case GuildPermission.ManageChannels:
                    case GuildPermission.ManageEmojis:
                    case GuildPermission.ManageGuild:
                    case GuildPermission.ManageRoles:
                    case GuildPermission.ManageWebhooks:
                    case GuildPermission.ViewAuditLog:
                    case GuildPermission.ManageNicknames:
                    case GuildPermission.ManageMessages:
                        score = score + 10;
                        continue;
                    case GuildPermission.BanMembers:
                    case GuildPermission.DeafenMembers:
                    case GuildPermission.KickMembers:
                    case GuildPermission.MoveMembers:
                    case GuildPermission.MuteMembers:
                    case GuildPermission.MentionEveryone:
                    case GuildPermission.SendTTSMessages:
                    case GuildPermission.PrioritySpeaker:
                        score = score + 5;
                        continue;
                    case GuildPermission.UseVAD:
                    case GuildPermission.UseExternalEmojis:
                    case GuildPermission.AttachFiles:
                    case GuildPermission.CreateInstantInvite:
                    case GuildPermission.AddReactions:
                    case GuildPermission.ChangeNickname:
                    case GuildPermission.EmbedLinks:
                    case GuildPermission.ReadMessageHistory:
                    case GuildPermission.Stream:
                        score = score + 2;
                        continue;
                    default:
                        score = score + 1;
                        continue;
                }
            }

            var percent = Math.Round((score / 142.00) * 100.00);
            int level;

            if (percent < 10.00) level = 0;
            else if (percent < 20.00) level = 1;
            else if (percent < 40.00) level = 2;
            else if (percent < 60.00) level = 3;
            else level = 4;

            return (percent, level, permList.Count);
        }

        [Command]
        public async Task WhoIsAsync(IUser user) 
        {
            var username = $"{user.Username}#{user.DiscriminatorValue}";
            var isHuman = user.IsBot | user.IsWebhook ? "Bot" : "User";
            var guilds = await Context.Client.GetGuildsAsync();
            var locatedOn = 0;

            var embedBuilder = new EmbedBuilder() 
            {
                Title = $"Who is {Format.Bold(username)}?",
                Color = Color.Blue,
                ThumbnailUrl = user.GetAvatarUrl(),
                Footer = new EmbedFooterBuilder() { Text = $"User ID : {user.Id}" }
            };

            foreach (var guild in guilds) if ((await guild.GetUserAsync(user.Id)) != null) locatedOn++;

            embedBuilder.AddField("User Type:", isHuman, true)
                .AddField("Registered:", user.CreatedAt, true)
                .AddField("Located On:", $"{locatedOn} servers", true);
            
            var gUser = await Context.Guild.GetUserAsync(user.Id);

            if (gUser != null) 
            {
                var roles = new StringBuilder();
                var result = GetEsperLevel(gUser);

                foreach (var roleId in gUser.RoleIds) roles.AppendFormat("{0} ", Context.Guild.GetRole(roleId).Mention);

                embedBuilder.AddField("Server Nickname:", gUser.Nickname ?? Format.Bold("N/A"), true)
                    .AddField("Joined Server At:", gUser.JoinedAt.ToString() ?? Format.Bold("Recently"), true)
                    .AddField("Current Server Roles:", roles.ToString() ?? Format.Bold("N/A"))
                    .AddField("ESPer Level:", $"{result.Level} ({result.Percent}% - {result.PermissionCount} Guild Permissions)");
            }

            await ReplyAsync(embed:embedBuilder.Build());
        }

        [Command]
        public async Task WhoIsAsync(ulong userId) 
        {
            var user = await Context.Client.GetUserAsync(userId);

            if (user == null)
            {
                await ReplyAsync("Can not find user.");
                return;
            }

            await WhoIsAsync(user);
        }

        [Command]
        public Task WhoIsAsync() => WhoIsAsync(Context.Author);
    }
}